src/stream.iced

```diff
@@ -0,0 +1,196 @@
+  _encrypt : (chunk, cb) =>
+    esc = make_esc(cb, "NaClEncryptStream::_encrypt")
+    # on the first call, spit out a header packet before writing payload
+    if not @_header_written
+      args = {
+        encryptor: @_encryptor,
+        recipients: @_recipients,
+        anonymized_recipients: @_anonymized_recipients
+      }
+      await header.generate_encryption_header_packet(args, esc(defer({header_intermediate, header_hash, mac_keys, payload_key})))
+      @_header_hash = header_hash
+      @_mac_keys = mac_keys
+      unless @_encryptor?
+        @_encryptor = nacl.alloc({force_js : false})
+      @_encryptor.secretKey = payload_key
+      @push(header_intermediate)
+      @_header_written = true
+    # either way, we want to encrypt the received payload
+    args = {
+      payload_encryptor: @_encryptor,
+      plaintext: chunk,
+      block_num: @_block_num,
+      header_hash: @_header_hash,
+      mac_keys: @_mac_keys
+    }
+    await payload.generate_encryption_payload_packet(args, esc(defer(payload_list)))
+    ++@_block_num
+    return cb(null, payload_list)
+
+  # write the empty payload packet
+  flush_append : (cb) ->
+    esc = make_esc(cb, "NaClEncryptStream::_flush_append")
+    await @_encrypt(new Buffer(''), esc(defer(payload_list)))
+    return cb(null, payload_list)
+
+  constructor : (@_encryptor, @_recipients, @_anonymized_recipients) ->
+    @_header_written = false
+    @_block_num = 0
+    @_mac_keys = null
+    @_header_hash = null
+    super({transform_func : @_encrypt, block_size : SALTPACK_BLOCK_LEN, readableObjectMode : true})
+
+# This is a raw NaCl stream - it doesn't implement msgpack, armoring, or anything like that. This should only ever be called inside the exposed EncryptStream class
+class NaClDecryptStream extends require('stream').Transform
+
+  # as above, this function encrypts for a specific key and is passed to `node-chunk-stream`
+  _decrypt : (chunk, cb) =>
+    esc = make_esc(cb, "NaClDecryptStream::_decrypt")
+    args = {
+      payload_decryptor: @_decryptor,
+      payload_list: chunk,
+      block_num: @_block_num,
+      header_hash: @_header_hash,
+      mac_key: @_mac_key,
+      recipient_index: @_recipient_index
+    }
+    await payload.parse_encryption_payload_packet(args, esc(defer(payload_text)))
+    ++@_block_num
+    return cb(null, payload_text)
+
+  _transform_chunk : (chunk, cb) =>
+    esc = make_esc(cb, "NaClDecryptStream::_transform_chunk")
+
+  # as above, this function parses the header on the first call and defers to `node-chunk-stream` on subsequent calls
+  _transform : (chunk, encoding, cb) ->
+    esc = make_esc(cb, "NaClDecryptStream::_transform")
+    if chunk.length is 0
+      return cb(null, null)
+    if @_header_read
+      await @_decrypt(chunk, esc(defer(out)))
+      if out.length is 0
+        @_found_empty_ending_packet = true
+      return cb(null, out)
+    else
```

Comment:
Conceptually, it makes more sense to me that the header-parsing code would come above the payload parsing code. Not a big deal if you prefer it this way though.